Added configuration screen

HenryLoenwind opened the pull request:
It's not completely finished, but very close to a state I'd stop with.

This re-adds a generic configuration UI that can show and edit all the values the configuration system can handle. It is opt-in, and mods can use it as a base class to do their own thing. It mostly uses vanilla's configuration system with a couple of tweaks on top.

There are no breaking changes. It is mostly self-contained, with the exception of an addition to how lists are configured. For one, it allows mods to restrict the lengths of lists (in both directions), but more importantly, it gives the UI a way of creating new elements to add to lists.

I also fixed 2 configuration values that still had "forge" in the code but "neoforge" in the language file.

Open TODOs:

- [x] Warning screen after changing values that require a world/client restart (important)
- [x] "Reset to default" button
- [x] "Undo" button
- [x] More javadoc
- [x] PR to documentation repo
- [ ] Narrator for custom widget (help wanted as tagged in the source)
- Edit dedicated server config (out of scope)
- Discarded: Visually mark changed values (vs default/vs original value)

To use this in a mod, add `container.registerExtensionPoint(IConfigScreenFactory.class, ConfigurationScreen::new);` to your client-mod class constructor. (Although this probably works fine in the normal mod class, but the advanced constructor wouldn't.)

![image](https://github.com/neoforged/NeoForge/assets/1485873/eb11a687-5b85-483a-8d5e-43b593c2ee35)
![image](https://github.com/neoforged/NeoForge/assets/1485873/011a84a4-fd7d-476d-804a-8e4c704e3f1e)
![image](https://github.com/neoforged/NeoForge/assets/1485873/0053aa10-304c-445d-8e60-e1a63652dd3c)
![image](https://github.com/neoforged/NeoForge/assets/1485873/655912f6-c19a-46ce-960b-fd2ec7e88885)
![image](https://github.com/neoforged/NeoForge/assets/1485873/48e46945-da16-4444-aa14-0f70a3041025)
![image](https://github.com/neoforged/NeoForge/assets/1485873/83c8783b-9b26-4aa0-8abb-fd1d5eac08c7)

HenryLoenwind committed:
Added configuration screen

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/signed)](https://cla-assistant.io/neoforged/NeoForge?pullRequest=1199) <br/>All committers have signed the CLA.

HenryLoenwind committed:
Added placeholders for weird config values

People put all kinds of stuff into the system, like List<List<Object>> = { {String, Number}, ... }. Better at least show the user that there is something or they might think we're broken.

HenryLoenwind committed:
Started preparing for undo

HenryLoenwind committed:
Implemented handling for changes that force game/server restart

HenryLoenwind commented:
Also, if someone would clean up neoforge's own configuration in regards to which ones are missing entries in the language file and which entries in the language file are zombies from old versions, that'd be great.

For this, note that the UI tries to get the tooltip from the language file and only uses the comment as fallback. While this doubles the text into two places, it also allows the tooltip to be translated and the comment to be present on servers.

IchHabeHunger54 commented:
Hello from the docs team!

The entire config article still needs a rework, so documentation for this system should best be added alongside that rework as a whole. In other words, don't let a missing docs PR stop you from merging this.

HenryLoenwind committed:
Finished undo system, fixed derped list screen logic

HenryLoenwind committed:
Added detection for "open to LAN" integrated servers

Can't edit values that have been synced to others or could be synced in the middle of editing.

HenryLoenwind committed:
Better propagate change from list screen

HenryLoenwind committed:
Added reset to default and started better undo system

or should I say "more sane undo system"?

HenryLoenwind committed:
Redid undo system

And prevent invalid lists from being committed

HenryLoenwind committed:
Cleanup pass

Including some minor changes for better extendibility

HenryLoenwind committed:
Merge remote-tracking branch 'neoforged/1.21.x' into ui

HenryLoenwind committed:
Removed debugging code

HenryLoenwind committed:
Fixed wrong list size check

HenryLoenwind committed:
Added documentation to and exposed sizeRange in ModConfigSpec

Also deprecated old methods that don't have a new value supplier

HenryLoenwind committed:
Merge remote-tracking branch 'neoforged/1.21.x' into ui

HenryLoenwind committed:
More javadoc and a tiny bit of cleanup

HenryLoenwind committed:
Merge remote-tracking branch 'neoforged/1.21.x' into ui

HenryLoenwind committed:
Found the better symbols for up/down/delete

TelepathicGrunt commented on src/main/java/net/neoforged/neoforge/client/ClientNeoForgeMod.java:
```diff
@@ -24,11 +27,13 @@
 
 @Mod(value = "neoforge", dist = Dist.CLIENT)
 public class ClientNeoForgeMod {
```
Could you slip in an internal annotation for this class? It's technically an internal class that isn't for modders to call

HenryLoenwind replied:
sure can do

HenryLoenwind committed:
Merge remote-tracking branch 'neoforged/1.21.x' into ui

HenryLoenwind committed:
Tagged neo's client mod as not being an API for mods...

Technici4n commented:
This can now be rebased on top of the latest config changes. ;)

HenryLoenwind committed:
Merge remote-tracking branch 'neoforged/1.21.x' into ui

HenryLoenwind committed:
Adapted to #1239

HenryLoenwind committed:
Cleaned up config-related translations

Removed unused keys and added missing ones

HenryLoenwind committed:
Missed a final applyFormatting run...

Matyrobbrt committed:
A bit of cleanup

Matyrobbrt committed:
Fix spotless stuff

Matyrobbrt committed:
Make the size range default to a signed int check

Matyrobbrt committed:
Some fallbacks

Matyrobbrt committed:
Breadcrumbs

Matyrobbrt committed:
Change ordering of CSF parameters

HenryLoenwind committed:
Removed extra sections

and made breadcrumb character translatable for RTL languages

Matyrobbrt committed:
Test config

HenryLoenwind committed:
Merge remote-tracking branch 'origin/ui' into ui

Matyrobbrt committed:
Merge branch 'ui' of https://github.com/HenryLoenwind/NeoForge into pr/1199

Matyrobbrt committed:
Breadcrumb the list path

Matyrobbrt committed:
Translate Enum values

HenryLoenwind committed:
Added more test configs

Fixed button overlap
Reinstated translation warning
Split translation warning
Fixed "single config" detection

HenryLoenwind committed:
Removed caption from silder

HenryLoenwind committed:
Added shallow tooltip and content based comment pre-processing

HenryLoenwind committed:
Merge remote-tracking branch 'neoforged/1.21.x' into ui

HenryLoenwind committed:
Merge remote-tracking branch 'neoforged/1.21.x' into ui

HenryLoenwind committed:
Fixed range not being added when using translated tooltips

HenryLoenwind committed:
Changed untranslated key listing to copy&paste-ready json

HenryLoenwind committed:
Added filename to tooltip

TelepathicGrunt commented:
I am approving this PR based off of end modder behavior and not the code portion. Just be sure to add to PR description that modders gets a console output of untranslated keys only if they visit the screens with the untranslated stuff and then exit the config gui. This should be stated so modders know this behavior exists.

Narrator support could be a future PR

TelepathicGrunt commented:
@HenryLoenwind Looks like there's a conflicting file rn

HenryLoenwind committed:
Merge branch '1.21.x' into ui

mezz commented on src/main/java/net/neoforged/neoforge/client/gui/IConfigScreenFactory.java:
```diff
@@ -24,7 +23,7 @@ public interface IConfigScreenFactory extends IExtensionPoint {
      * Creates a new config screen. The {@code modListScreen} parameter can be used for a "back" button.
      */
-    Screen createScreen(Minecraft minecraft, Screen modListScreen);
+    Screen createScreen(ModContainer container, Screen modListScreen);
```
@HenryLoenwind This is a (small) breaking change, please include the breaking change tag on this commit for visibility or add a backward compatible method.

HenryLoenwind replied:
@Matyrobbrt Your change ^
